### PR TITLE
ringbuf cleanup

### LINKED ISSUE #6815: Incorporate ringbuf improvements from #6214
#6214 for 7.3.x was closed without merging, but contains a cleanup of `ringbuf` and its callers. Redo this for `main`.

### PR DESCRIPTION
**Pull request by dhalbert**
- Fixes #6815 

Derived from #6124, with further cleanup:

- Now use `ringbuf_init()` in various places instead of reaching inside the ringbuf struct.
- ringbuf no longer has the "capacity is one less than size" oddity. It used an extra slot to indicate full and empty. Now it uses a count. Sometimes this technique is used when doing atomic operations, but we already don't expect it to be atomic. Added a comment to that effect. Changed `ringbuf_capacity()` to `ringbuf_size()` to simplify naming.
- `ringbuf_free()` started to rely on gc to free its buffer in #6800. Since it now doesn't really free, the name was changed to `ringbuf_deinit()`. Removed some flags about whether a ringbuf buffer is heap-allocated or not, because now it doesn't matter.
- Improved internal names.
- Removed 16-bit peek routine, which was not used anywhere.
- Fixed up unix coverage test of ringbuf.

Tested by:
- connecting two RP2040 UARTs together and sending longer and longer byte strings
- nRF BLE heart rate monitor, to test PacketBuffer.
- BLE UART test

### CODE REVIEW

File: ports/broadcom/common-hal/busio/UART.c [hunk 1]
```diff
@@ -208,7 +208,9 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
     self->sigint_enabled = sigint_enabled;
 
     if (rx != NULL) {
+        // Use the provided buffer when given.
         if (receiver_buffer != NULL) {
+            ringbuf_init(&self->ringbuf, receiver_buffer, receiver_buffer_size);
             self->ringbuf = (ringbuf_t) { receiver_buffer, receiver_buffer_size };
```
Comment (jepler) [inline ports/broadcom/common-hal/busio/UART.c:214]: looks funky

**Review by jepler — CHANGES_REQUESTED** [review]

File: ports/broadcom/common-hal/busio/UART.c (hunk 1 above)
Comment (dhalbert) [reply ports/broadcom/common-hal/busio/UART.c]: Bad edit, and fixed another issue too. Thanks.

**Review by jepler — APPROVED** [review]
I didn't spot any further problems, but I didn't test either.

**Comment by tlyu** [comment]
I tested the pull request build artifacts with the `ble_adafruit_clue.py` example on a Clue and confirmed that the Tone service now works from the iOS Bluefruit Playground client. (It did not work with previous CircuitPython 8.x builds.)

**Comment by dhalbert** [comment]
Hmm - I thought I had merged #6800 (which was added to 7.3.x) into `main`, but I maybe I didn't? Anway, glad it fixed the Tone Service in 8.0.0 as well.

**Pull request merged**